Changes to support `torch._export.aot_compile`

agunapal opened the pull request:
## Description

This PR 
- Adds support for `torch._export.aot_compile`
- Includes an example with ResNet18 with `max-autotune` and `dynamic_shapes`
- Tested equivalence with multiple loads

Comparison with `torch.compile`

<google-sheets-html-origin><style type="text/css"><!--td {border: 1px solid #cccccc;}br {mso-data-placement:same-cell;}--></style>
Model | Mode | Model loading(ms) | First Inference Time (ms)
-- | -- | -- | --
  |   |   |  
ResNet 18 | compile + max-autotune | 4111 | 25918
VGG 16 | compile + max-autotune | 5284 | 64960
ResNet 18 | torch._export.aot_compile + max-autotune | 15876 | 2704
VGG 16 | torch._export.aot_compile + max-autotune | 16042 | 2794




Fixes #(issue)

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] This change requires a documentation update

## Feature/Issue validation/testing

```
pytest -v test_torch_export.py 
================================================================================== test session starts ===================================================================================
platform linux -- Python 3.10.13, pytest-7.3.1, pluggy-1.0.0 -- /home/ubuntu/anaconda3/envs/ts_sam_Dec1/bin/python
cachedir: .pytest_cache
rootdir: /home/ubuntu/serve
plugins: cov-4.1.0, mock-3.12.0
collected 1 item                                                                                                                                                                         

test_torch_export.py::test_torch_export_aot_compile PASSED                                                                                                                         [100%]

==================================================================================== warnings summary ====================================================================================
test_torch_export.py:4
  /home/ubuntu/serve/test/pytest/test_torch_export.py:4: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google.logging')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2348
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2348: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(parent)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/pkg_resources/__init__.py:2868: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('ruamel')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../../../anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/transformers/utils/generic.py:441
  /home/ubuntu/anaconda3/envs/ts_sam_Dec1/lib/python3.10/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
    _torch_pytree._register_pytree_node(

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
============================================================================= 1 passed, 7 warnings in 45.19s =============================================================================
(ts_sam_Dec1) ubuntu@ip-172-31-11-40:~/serve/test/pytest$ 

```

Checking Equivalance when loading multiple times

Used the following code
```
import pickle
import torch

FILE1 = "test_load.pkl"
FILE2 = "test_load1.pkl"
print("Reading output 1")
with open(FILE1,'rb') as f:
    output1 = pickle.load(f)

print("Reading output 2")
with open(FILE2,'rb') as f:
    output2 = pickle.load(f)


print("Checking if output1 == output2", torch.equal(output1, output2))
```
results in 

```
python test_equivalance.py 
Reading output 1
Reading output 2
Checking if output1 == output2 True

```

- Model Loading logs
```
2023-12-13T20:03:15,259 [INFO ] W-9000-res18-pt2_1.0-stdout MODEL_LOG - torch._export is an experimental feature! Succesfully loaded torch exported model.
2023-12-13T20:03:15,259 [INFO ] W-9000-res18-pt2_1.0-stdout MODEL_LOG - export is not a supported backend
2023-12-13T20:03:15,264 [INFO ] W-9000-res18-pt2_1.0 org.pytorch.serve.wlm.WorkerThread - Backend response time: 17378
2023-12-13T20:03:15,264 [DEBUG] W-9000-res18-pt2_1.0 org.pytorch.serve.wlm.WorkerThread - W-9000-res18-pt2_1.0 State change WORKER_STARTED -> WORKER_MODEL_LOADED
2023-12-13T20:03:15,265 [INFO ] W-9000-res18-pt2_1.0 TS_METRICS - WorkerLoadTime.Milliseconds:19913.0|#WorkerName:W-9000-res18-pt2_1.0,Level:Host|#hostname:ip-172-31-11-40,timestamp:1702497795
2023-12-13T20:03:15,265 [INFO ] W-9000-res18-pt2_1.0 TS_METRICS - WorkerThreadTime.Milliseconds:3.0|#Level:Host|#hostname:ip-172-31-11-40,timestamp:1702497795
```

- Model loaded on 4 GPUs
![Screenshot 2023-12-18 at 3 29 31 PM](https://github.com/pytorch/serve/assets/16617092/627e1b05-28ac-4be0-9104-7f5b735df189)


## Checklist:

- [ ] Did you have fun?
- [ ] Have you added tests that prove your fix is effective or that this feature works?
- [ ] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

chauhang commented:
@ankithagunapal Thanks for adding this support. Why is the model load taking so longer with AOTCompile?

agunapal commented:
> @ankithagunapal Thanks for adding this support. Why is the model load taking so longer with AOTCompile?

@chauhang Yet to run the profiler on the loading of the .so file. Will check and follow-up

msaroufim commented on examples/pt2/README.md:
```diff
@@ -107,3 +107,22 @@ print(extra_files['foo.txt'])
+
+## torch._export.aot_compile
+
+Using `torch.compile` to wrap your existing eager PyTorch model can result in out of the box speedups. However, `torch.compile` is a JIT compiler. TorchServe has been supporting `torch.compile` since PyTorch 2.0 release. In a production setting, when you have multiple instances of TorchServe, each of of your instances would `torch.compile` the model on inference. TorchServe's model archiever is n ...
```
```suggestion
Using `torch.compile` to wrap your existing eager PyTorch model can result in out of the box speedups. However, `torch.compile` is a JIT compiler. TorchServe has been supporting `torch.compile` since PyTorch 2.0 release. In a production setting, when you have multiple instances of TorchServe, each of of your instances would `torch.compile` the model on the first inference. TorchServe's model archiver is not able to truly guarantee reproducibility because its a JIT compiler.
```

msaroufim commented on examples/pt2/README.md:
```diff
@@ -107,3 +107,22 @@ print(extra_files['foo.txt'])
+
+In addition, the first inference request with `torch.compile` will be slow as the model needs to compile.
+
+To solve this problem, `torch.export` has an experimental API `torch._export.aot_compile` which is able to `torch.export` a torch compilable model with no graphbreaks along with AOTInductor.
```
```suggestion
To solve this problem, `torch.export` has an experimental API `torch._export.aot_compile` which is able to `torch.export` a torch compilable model if it has no graph breaks and then run it with AOTInductor
```

msaroufim commented on examples/pt2/torch_export_aot_compile/README.md:
```diff
@@ -0,0 +1,69 @@
+# TorchServe inference with torch._export aot_compile
```
duplicate docs?

agunapal replied:
resolved. Thanks

msaroufim commented on examples/pt2/torch_export_aot_compile/README.md:
```diff
@@ -0,0 +1,69 @@
+- `CUDA 12.1`
+
+Change directory to the examples directory
+Ex:  `cd` to `examples/pt2/torch_export_aot_compile`
```
```suggestion
Ex:  `cd  examples/pt2/torch_export_aot_compile`
```

msaroufim commented on examples/pt2/torch_export_aot_compile/README.md:
````diff
@@ -0,0 +1,69 @@
+
+Install PyTorch 2.2 nightlies by running
+```
+chmod +x install_pytorch_nightlies.sh
````
we don't need this script, install_dependencies.py has a nightly flag

msaroufim commented on examples/pt2/torch_export_aot_compile/README.md:
```diff
@@ -0,0 +1,69 @@
+
+The model is saved with `.so` extension
+Here we are torch exporting with AOT Inductor with `max_auotune` mode.
+This is also making use of `dynamic_shapes` to support batch size from 1 to 32.
```
did this interact well with dynamic batching? What happens if you set the batch size to something really large instead?

agunapal replied:
tried with batch_size 16. It works. I guess we launch it as an experimental feature and then based on issues/feedback, address them

msaroufim commented on examples/pt2/torch_export_aot_compile/README.md:
```diff
@@ -0,0 +1,69 @@
+The model is saved with `.so` extension
+Here we are torch exporting with AOT Inductor with `max_auotune` mode.
+This is also making use of `dynamic_shapes` to support batch size from 1 to 32.
+In the code, the min batch_size is mentioned as 2 instead of 1. You can find an explanation for this [here](https://pytorch.org/docs/main/export.html#expressing-dynamism)
```
can you bring core idea? Batch size 1 is what people choose if they want low latency

agunapal replied:
I read the doc , but I don't understand it tbh

msaroufim commented on requirements/common_gpu.txt:
```diff
@@ -1,2 +1,3 @@
 nvgpu; sys_platform != 'win32'
 nvgpu==0.10.0; sys_platform == 'win32'
+ninja==1.11.1.1
```
is ninja needed to run an inference or to export? If export only we shouldn't add this as a dependency

agunapal replied:
This is needed for loading the .so file. I can add this in requirements.txt for the example?

agunapal replied:
This doesn't work.

msaroufim commented on ts/handler_utils/torch_export/load_model.py:
```diff
@@ -0,0 +1,27 @@
+
+def load_exported_model(model_so_path, device):
+    module = load_inline(
+        name="aot_inductor",
```
would we need to change the name in the case of multiple workers?

agunapal replied:
tried with multiple workers. Shared a screenshot of nvidia-smi in the PR

msaroufim commented on ts/handler_utils/torch_export/load_model.py:
```diff
@@ -0,0 +1,27 @@
+        name="aot_inductor",
+        cpp_sources=[aot_inductor_launcher(model_so_path, device)],
+        # use a unique build directory to avoid test interference
+        build_directory=tempfile.mkdtemp(dir=cache_dir()),
```
does this use the same cache_dir as teh environment variable a user would be setting?

agunapal replied:
This seems like `TORCHINDUCTOR_CACHE_DIR`

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -151,6 +155,12 @@ def initialize(self, context):
             self.map_location = "cpu"
             self.device = torch.device(self.map_location)
 
+        TORCH_EXPORT_AVAILABLE = False
```
set these global variables in packaging if condition instead

agunapal replied:
I looked at the logic again. This seems like the best place.

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -53,6 +53,10 @@
     )
     PT2_AVAILABLE = False
 
+if packaging.version.parse(torch.__version__) > packaging.version.parse("2.1.1"):
```
i think packaging throws a bunch of warnings right? Can we do whatever the warning recommends cause our logs are starting to get noisy

agunapal replied:
I don't see any warnings from packaging

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -53,6 +53,10 @@
     PT2_AVAILABLE = False
 
+if packaging.version.parse(torch.__version__) > packaging.version.parse("2.1.1"):
+    PT220_AVAILABLE = True
```
does this mean pt.2.2? Can we call this call PT2_2-0 should be easier to read

agunapal replied:
Doesn't allow `PT2_2-0`

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -180,6 +190,13 @@ def initialize(self, context):
             self.model = setup_ort_session(self.model_pt_path, self.map_location)
             logger.info("Succesfully setup ort session")
 
+        elif self.model_pt_path.endswith(".so") and TORCH_EXPORT_AVAILABLE:
```
archiver doc strings need an update too to make it clear it also supports .so file

agunapal replied:
not sure which one you are specifically referring to?
This doesn't talk about the format of the file
```
usage: torch-model-archiver [-h] --model-name MODEL_NAME [--serialized-file SERIALIZED_FILE] [--model-file MODEL_FILE] --handler HANDLER [--extra-files EXTRA_FILES]
                            [--runtime {python,python3}] [--export-path EXPORT_PATH] [--archive-format {tgz,no-archive,zip-store,default}] [-f] -v VERSION [-r REQUIREMENTS_FILE]
                            [-c CONFIG_FILE]
torch-model-archiver: error: the following arguments are required: --model-name, --handler, -v/--version
```

msaroufim reviewed (changes requested):
Looks mostly good, a few minor things left to do - also still not super clear on how the timing measurements were obtained - I'd add some log.info statements

agunapal commented:
> Looks mostly good, a few minor things left to do - also still not super clear on how the timing measurements were obtained - I'd add some log.info statements

@msaroufim Thanks. I have addressed most of the feedback. However, wondering how we should address support for CPU. Does aot_compile on CPU make sense?

agunapal commented on test/pytest/test_torch_export.py:
```diff
@@ -0,0 +1,88 @@
+    os.chdir(tmp_path)
+
+
+@pytest.mark.skipif(PT_220_AVAILABLE == False, reason="torch version is < 2.2.0")
```
added. Thanks

msaroufim commented:
CPU support is a completely valid scenario, inductor codegens native c++ that the intel team has been optimizing

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -151,6 +155,12 @@ def initialize(self, context):
+        if hasattr(self, "model_yaml_config") and "pt2" in self.model_yaml_config:
+            pt2_value = self.model_yaml_config["pt2"]
+            if pt2_value == "export" and PT220_AVAILABLE:
+                USE_TORCH_EXPORT = True
```
why do you need this bool flag?

msaroufim commented on ts/torch_handler/base_handler.py:
```diff
@@ -180,6 +190,15 @@ def initialize(self, context):
+        elif self.model_pt_path.endswith(".so") and USE_TORCH_EXPORT:
+            # Set cuda stream to the gpu_id of the backend worker
+            if torch.cuda.is_available() and properties.get("gpu_id") is not None:
+                torch.cuda.set_stream(torch.cuda.Stream(int(properties.get("gpu_id"))))
```
could you add some comment here? Why are we launching things on specific streams?

msaroufim reviewed (approved):
stamping to unblock but this needs a few more things before merge
1.  This should indeed work with cpu so add a test
2. Get rid of the install nightlies script and instead use the good old install_dependencies.py script
3. Add a comment on the cuda streams section
4. You don't need the export bool flag IMO

agunapal commented:
> stamping to unblock but this needs a few more things before merge
> 
> 1. This should indeed work with cpu so add a test
> 2. Get rid of the install nightlies script and instead use the good old install_dependencies.py script
> 3. Add a comment on the cuda streams section
> 4. You don't need this flag  USE_TORCH_EXPORT

1. Tested on CPU. pytest runs on both CPU/GPU. There is a limitation for batch_size on CPU. Mentioned this in the script.
2. I have updated instructions to have both. Most users using torchserve already have dependencies installed. the torch version doesn't get overwritten. It needs to be uninstalled and then installed again. Hence, script is needed
3. Good catch. Turns out this was not needed. We just need torch.cuda.set_device
4. Refactored the code. 

Also, changed the config to the following. (in case they add other options in export in the future)
```
pt2 :
  export:
    aot_compile: true
```
